### PR TITLE
Add a requires_grad_() function to tensors.

### PR DESCRIPTION
**Pull request by gchanan**

### CODE REVIEW

**Comment by gchanan** [comment]
This is meant to be a replacement of https://github.com/pytorch/pytorch/pull/6740, since that is unsafe is certain conditions (if it's done multiple times).

**Review by apaszke — APPROVED** [review]

File: tools/autograd/templates/python_variable_methods.cpp [hunk 1]
```diff
@@ -408,6 +409,26 @@ static PyObject * THPVariable_record_stream(PyObject* self, PyObject* arg)
...
+  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
+  ParsedArgs<1> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+  auto requires_grad = r.toBool(0);
+  // should we throw if requires_grad is true?  var.requires_grad = True throws here
+  // but it's nice to let this be a no-op.
+  if (!self_.is_leaf() && !requires_grad) {
+    throw std::runtime_error(autograd::utils::requires_grad_leaf_error(requires_grad));
+  }
+  self_.set_requires_grad(requires_grad);
```
Comment (SsnL) [inline tools/autograd/templates/python_variable_methods.cpp]: Wouldn't this throw the above error anyways?

Comment (gchanan) [reply tools/autograd/templates/python_variable_methods.cpp]: no, the c++ variable version doesn't check; maybe we should add a "safe" version for the future.

Comment (SsnL) [reply tools/autograd/templates/python_variable_methods.cpp]: Oh I see. You are right.

**Pull request merged**